components/scream/src/share/scream_kokkos_meta.hpp

```diff
@@ -0,0 +1,38 @@
+#ifndef INCLUDE_SCREAM_KOKKOS_META
+#define INCLUDE_SCREAM_KOKKOS_META
+
+#include <Kokkos_Core.hpp>
+
+namespace scream {
+namespace ko {
+
+/* Kokkos-related types not specific to the app. Thus, do not include
+   app-specific array types here; rather, include only those things that one
+   could imagine could go into Kokkos itself.
+*/
+
+// Make the input View Unmanaged, whether or not it already is. One might
+// imagine that View::unmanaged_type would provide this.
+//   Use: typename Unmanged<ViewType>::type
+template <typename View>
+struct Unmanaged {
+  // Provide a full View type specification, augmented with Unmanaged.
+  typedef Kokkos::View<typename View::traits::scalar_array_type,
+                       typename View::traits::array_layout,
+                       typename View::traits::device_type,
+                       Kokkos::MemoryTraits<
+                         // All the current values...
+                         (View::traits::memory_traits::RandomAccess ? Kokkos::RandomAccess : 0) |
+                         (View::traits::memory_traits::Atomic ? Kokkos::Atomic : 0) |
+                         (View::traits::memory_traits::Restrict ? Kokkos::Restrict : 0) |
+                         (View::traits::memory_traits::Aligned ? Kokkos::Aligned : 0) |
+                         // ... |ed with the one we want, whether or not it's
+                         // already there.
+                         Kokkos::Unmanaged> >
```

Comment:
Wouldn't `Kokkos::MemoryTraits<View::traits::memory_traits | Kokkos::Unmanaged>` do the trick, without the need of listing all the current memory traits? I do not know how likely it is that Kokkos will add more traits, or remove/rename some existing ones, but still.

Reply:
Good question. The simple answer is that can't work. But the longer one is that Kokkos I think _could_ support such behavior.

Simple answer: `memory_traits` is an alias for `Kokkos::MemoryTraits`, which is a struct of `enum : bool`s. In particular, `Kokkos::View` appears not to have any record of the underlying `MemoryTraitsFlag` unsigned int mask that is a template parameter to `MemoryTraits`. Thus, there is nothing with which to |.

Longer one: But one could imagine adding `enum : unsigned { mask = T };` to `Kokkos::MemoryTraits`, thus making `View::traits::memory_traits::mask | Kokkos::Unmanaged` possible in user code.

The second answer is the reason I want a separate scream_kokkos_meta header that has completely app-unspecific things. These could in principle go away with changes to Kokkos, so we should isolate such things.

Reply:
The nice thing is that even though the impl is ugly, the usage is super simple: `Unmanaged<ViewType>`. IMO, this is better than hard-typed unmanaged variants of every app-specific View typedef. `Unmanaged<View>::non_const_type` should be another common thing we use to have good const correctness in kernels.